compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -1352,6 +1497,60 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             return Ok(None);
         }
 
+        let Some((def_id, args)) = self.probe_inherent_assoc_shared(
+            segment,
+            adt_did,
+            self_ty,
+            block,
+            span,
+            ty::AssocKind::Type,
+        )?
+        else {
+            return Ok(None);
+        };
+
+        let ty = Ty::new_alias(tcx, ty::Inherent, ty::AliasTy::new_from_args(tcx, def_id, args));
+        Ok(Some((ty, def_id)))
+    }
+
+    fn probe_inherent_assoc_const(
+        &self,
+        segment: &hir::PathSegment<'tcx>,
+        adt_did: DefId,
+        self_ty: Ty<'tcx>,
+        block: HirId,
+        span: Span,
+    ) -> Result<Option<(Const<'tcx>, DefId)>, ErrorGuaranteed> {
+        let tcx = self.tcx();
+
+        let Some((def_id, args)) = self.probe_inherent_assoc_shared(
+            segment,
+            adt_did,
+            self_ty,
+            block,
+            span,
+            ty::AssocKind::Const,
+        )?
+        else {
+            return Ok(None);
+        };
+
+        let ct = Const::new_unevaluated(tcx, ty::UnevaluatedConst::new(def_id, args));
+        Ok(Some((ct, def_id)))
+    }
+
+    fn probe_inherent_assoc_shared(
+        &self,
+        segment: &hir::PathSegment<'tcx>,
+        adt_did: DefId,
+        self_ty: Ty<'tcx>,
+        block: HirId,
+        span: Span,
+        kind: ty::AssocKind,
+    ) -> Result<Option<(DefId, GenericArgsRef<'tcx>)>, ErrorGuaranteed> {
+        let tcx = self.tcx();
+
+        let name = segment.ident;
         let candidates: Vec<_> = tcx
             .inherent_impls(adt_did)
             .iter()
```

Comment:
the `filter_map` below this shouldn't hard code `AssocKind::Type` anymore

(sigh @ github why wont you let me leave reviews on code that is too far away from a diff...)